### ISSUE TITLE
GroupOperation - User Initiated

### ISSUE THREAD

**Issue by ryanjm** [danthorpe/Operations#263]
I just added this to one of a `GroupOperation` in my project and I was wondering you thought it would make sense in the project itself:

```
    override var userInitiated: Bool {
        get {
            return super.userInitiated
        }
        set {
            for operation in operations as! [Operation] {
                operation.userInitiated = newValue
            }
            super.userInitiated = newValue
        }
    }
```

The reasoning behind this is that often times in a group operation your `init` method will create many additional operations to go into the group. But you can't set the `userInitiated` property until later and you'll usually want that to apply to all operations in the group.

Example: I have a group operation for downloading a list of files. I have another operation, `DownloadFileOp`, which downloads one file so the group operation creates a lot of these. `DownloadFileOp` has a conditional that says to only download if on Wifi OR the user initiated the download. Therefore I want those operations to have their `userInitiated` property updated when the group operation's is set.

Is there a better / more appropriate way of handling this? Is it better to create a connivence `init` that is used to set `userInitiated` before the other operations are created?

If this was to be added to the project I think it'd need to handle the following:

- If additional operations are added to the queue later then they should get the same `userInitiated` value.
- The loop above should handle the case that some of the operations might not be of `Operation` type.

**Comment by ryanjm** [comment]
I tried the `convince init` route and the compiler won't let me due to:

> Use of 'self' in property access 'userInitiated' before self.initialized self.

Here was my code

```swift
convenience init(attachments: [Attachment], userInitiated ut: Bool) {
    userInitiated = ut
    self.init(attachments: attachments)
}
```

**Comment by danthorpe** [comment]
Hi @ryanjm - that for the feedback, quite a subtle yet interesting point here.

First up, yep, I definitely think that `GroupOperation` should set the *quality of service* for its child operations to match its own - which would require an override similar to what you propose here. 

But stepping back a bit... this `userInitiated` property is directly from @davedelong's sample code from WWDC. At the moment, it sort of conflates the concept of *intent* with the quality of service (which is a mechanism to indicate to the queue the priority of the block). And thinking about this - I'm not a huge fan.

So I propose doing something like this instead:

```swift
enum UserIntent {
    case Implicit, Explicit

    var qualityOfService: NSQualityOfService {
        switch self {
        case .Implicit: return .Default
        case .Explicit: return .UserInitiated
        }
    }
}

class Operation: NSOperation {
   // etc

    var userIntent: UserIntent = .Implicit {
        didSet {
            qualityOfService = userIntent.qualityOfService
        }
    }
}
```

Inside `GroupOperation`, I think that the `userIntent` property of the group should be inferred from the `userIntent` property of any `Operation` instances that it is initialized with. And that setting the `userIntent` or `qualityOfService` properties on the `GroupOperation` filters down to child operation *if the group has not started*.

I'm not sure about operations which are added to the group after the group has already started (but not yet finished). In all likelihood, this will be happening with a `GroupOperation` subclass - and it should be managing its own added child operations, so I'm leaning towards not overriding any `userIntent` or `qualityOfService` properties.

Thoughts?

**Comment by ryanjm** [comment]
I think the concept of adding `UserIntent` does make the code clearer and easier to understand.

As for adding that to any operation that is added once the group has started, I lean towards it needing to set `userIntent` on those as well. The group encompasses the _work_ expected to be done. If the user initiated it in the beginning, then it should be clear to every related operation that the user initiated it.

This would be very similar to the [log severity](https://github.com/danthorpe/Operations/blob/968fcefdb5f9a61bbf5368fa9db1b715aaf0a8dd/Sources/Core/Shared/GroupOperation.swift#L107) that is already copied to any added operation, right?

My example case would be that I have an object that I want to update, lets call it a `Post`. I don't know anything about the post's comments until I've downloaded it, therefore in the `willFinishOperation(_:withErrors:)` method I create additional download operations for each of the comments that are in that post. Those new operations should know about the same level of `userIntent`.

That said, it would be easy for me to add one line to `willFinishOperation` to add the `userIntent` to those new operations. Therefore, I guess it becomes more of an issue of documentation on whether new operations will inherit the the `userIntent` or not.

**Comment by jshier** [comment]
How would this setting affect operations running on queues that have an underlying `dispatch_queue_t` set, which has it's own QoS? I can't find much in the way of documentation about how the two interact.

**Comment by ryanjm** [comment]
@jshier I'm not sure. Why are you messing with `dispatch_queue_t`? `Operation`s / `NSOperation`s should be running on `NSOperationQueue`s. You can set the QoS on those but according to the [documentation](https://developer.apple.com/library/mac/documentation/Cocoa/Reference/NSOperationQueue_class/#//apple_ref/occ/instp/NSOperationQueue/qualityOfService) it'll be ignored if the operation's QoS is set:

> This property specifies the service level applied to operation objects added to the queue. If the operation object has an explicit service level set, that value is used instead. The default value of this property depends on how you created the queue. For queues you create yourself, the default value is NSOperationQualityOfServiceBackground. For the queue returned by the mainQueue method, the default value is NSOperationQualityOfServiceUserInteractive and cannot be changed.

Does that help answer your question?

**Comment by jshier** [comment]
No, since I'm talking about setting the `OperationQueue`s `underlyingQueue` to a particular, created queue, so that it's available to the running operations. However, I would think the documentation you cited would cover that case.

**Comment by ryanjm** [comment]
Then I'd think the same situation applies. I'd think setting the `NSOperationQueue`'s QoS would set the `underlyingQueue`'s QoS. Though I guess that would be something you'd have to test out. Have you found different results? What issue are you running into?

This particular issue is just a mater of how to set the operation's user intent and thus, effect the `qualityOfService` of the operation.

**Comment by danthorpe** [comment]
@jshier I don't know for sure how `NSOperationQueue` interprets this. I need to read the documentation - good find @ryanjm. But also I guess need to find out how `dispatch_queue_t` also works when it has a target queue. My instinct says that `NSOperationQueue` would select a new target queue with the corresponding QoS from the global pool, but only if the underlyingQueue is not explicitly set? But that's pure conjecture.

This is a little bit of a detour on the detail though. In general, separating out the concept of *intent* from *QoS* is :+1:? I can code it up and put a PR together to continue the discussion.

**Comment by ryanjm** [comment]
> separating out the concept of intent from QoS is :+1:?

:+1: Thanks.

**Comment by danthorpe** [comment]
Gonna close this, as PR merged. Will be released in 2.10
